File: demo/server/Judger.py
from server.game_pb2 import ScoreType


class Judger(object):
    def __init__(self):
        self.func = {
            ScoreType.Ones: self.Ones,
            ScoreType.Twos: self.Twos,
            ScoreType.Threes: self.Threes,
            ScoreType.Fours: self.Fours,
            ScoreType.Fives: self.Fives,
            ScoreType.Sixes: self.Sixes,
            ScoreType.Chance: self.Chance,
            ScoreType.FourOfAKind: self.FourOfAKind,
            ScoreType.FullHouse: self.FullHouse,
            ScoreType.SmallStraight: self.SmallStraight,
            ScoreType.LargeStraight: self.LargeStraight,
            ScoreType.Yahtzee: self.Yahtzee,
        }

    def judge(self, dices, selected_type):
        if len(dices) != 5 or not all(x >= 1 and x <= 6 for x in dices):
            raise Exception(
                "dices must be a 5-len array and all elements must be between 1-6")
        if selected_type not in self.func:
            raise Exception(
                "Invailed Choice")
        return self.func[selected_type](dices)

    def Ones(self, dices):
        return sum(x for x in dices if x == 1)

    def Twos(self, dices):
        return sum(x for x in dices if x == 2)

    def Threes(self, dices):
        return sum(x for x in dices if x == 3)

    def Fours(self, dices):
        return sum(x for x in dices if x == 4)

    def Fives(self, dices):
        return sum(x for x in dices if x == 5)

    def Sixes(self, dices):
        return sum(x for x in dices if x == 6)

    def Chance(self, dices):
        return sum(dices)

    def FourOfAKind(self, dices):  # For AAABB
        m = {}
        for x in dices:
            if x in m:
                m[x] += 1
            else:
                m[x] = 1
        for key in m:
            if m[key] >= 4:
                return sum(dices)
        return 0

    def FullHouse(self, dices):  # For AAABB
        m = {}
        for x in dices:
            if x in m:
                m[x] += 1
            else:
                m[x] = 1
        if len(m) == 1:  # Yahtzee is also FullHouse
            return sum(dices)
        if len(m) == 2:
            first_count = next(iter(m.values()))
            if first_count == 2 or first_count == 3:
                return sum(dices)
        return 0

    def SmallStraight(self, dices):  # For 1234,2345,3456
        patterns = [[1, 2, 3, 4], [2, 3, 4, 5], [3, 4, 5, 6]]
        for pattern in patterns:
            for x in pattern:
                if x not in dices:
                    break
            else:
                return 15  # TODO: hard code
        return 0

    def LargeStraight(self, dices):  # For 12345,23456
        patterns = [[1, 2, 3, 4, 5], [2, 3, 4, 5, 6]]
        for pattern in patterns:
            for x in pattern:
                if x not in dices:
                    break
            else:
                return 30  # TODO: hard code
        return 0

    def Yahtzee(self, dices):  # For AAAAA
        m = {}
        for x in dices:
            if x in m:
                m[x] += 1
            else:
                m[x] = 1
        if len(m) == 1:
            return 50  # TODO: hard code
        return 0

File: final/Judger.cpp

#include "Judger.h"

#include <map>
#include <numeric>
#include <set>
#include <algorithm>
#include <cassert>

int Judger::result[14][65536];


Judger::Judger()
{
	methods[ScoreType::Ones] = std::bind(&Judger::Ones, this, std::placeholders::_1);
	methods[ScoreType::Twos] = std::bind(&Judger::Twos, this, std::placeholders::_1);
	methods[ScoreType::Threes] = std::bind(&Judger::Threes, this, std::placeholders::_1);
	methods[ScoreType::Fours] = std::bind(&Judger::Fours, this, std::placeholders::_1);
	methods[ScoreType::Fives] = std::bind(&Judger::Fives, this, std::placeholders::_1);
	methods[ScoreType::Sixes] = std::bind(&Judger::Sixes, this, std::placeholders::_1);

	methods[ScoreType::Chance] = std::bind(&Judger::Chance, this, std::placeholders::_1);
	methods[ScoreType::ThreeOfAKind] = std::bind(&Judger::ThreeOfAKind, this, std::placeholders::_1);
	methods[ScoreType::FourOfAKind] = std::bind(&Judger::FourOfAKind, this, std::placeholders::_1);
	methods[ScoreType::FullHouse] = std::bind(&Judger::FullHouse, this, std::placeholders::_1);
	methods[ScoreType::SmallStraight] = std::bind(&Judger::SmallStraight, this, std::placeholders::_1);
	methods[ScoreType::LargeStraight] = std::bind(&Judger::LargeStraight, this, std::placeholders::_1);
	methods[ScoreType::Yahtzee] = std::bind(&Judger::Yahtzee, this, std::placeholders::_1);

	static bool inited = false;
	if (!inited)
		memset(result, -1, sizeof(result));
	inited = true;
}

int Judger::Judge(const YahtzeeStatus& status, ScoreType type)
{
	assert(status.dices.ready());

	compressed_dicehash_t dicehash = YahtzeeStatus::composeDices(status.dices);

	return JudgeByDicehash(dicehash, type);
}

int Judger::JudgeByDicehash(compressed_dicehash_t hash, ScoreType type)
{
	int bit = 0;
	while (((int)type & (1 << bit)) == 0 && bit < 14)
	{
		bit++;
	}
	if (result[bit][hash] != -1)
	{
		return result[bit][hash];
	}

	Dices dices = YahtzeeStatus::parseDices(hash);
	int score = methods[type](dices);
	return result[bit][hash] = score;
}

int Judger::Ones(const Dices& dices)
{
	int score = 0;
	for (int dice : dices.dices)
	{
		if (dice == 1)
		{
			score += dice;
		}
	}
	return score;
}

int Judger::Twos(const Dices& dices)
{
	int score = 0;
	for (int dice : dices.dices)
	{
		if (dice == 2)
		{
			score += dice;
		}
	}
	return score;
}

int Judger::Threes(const Dices& dices)
{
	int score = 0;
	for (int dice : dices.dices)
	{
		if (dice == 3)
		{
			score += dice;
		}
	}
	return score;
}

int Judger::Fours(const Dices& dices)
{
	int score = 0;
	for (int dice : dices.dices)
	{
		if (dice == 4)
		{
			score += dice;
		}
	}
	return score;
}

int Judger::Fives(const Dices& dices)
{
	int score = 0;
	for (int dice : dices.dices)
	{
		if (dice == 5)
		{
			score += dice;
		}
	}
	return score;
}

int Judger::Sixes(const Dices& dices)
{
	int score = 0;
	for (int dice : dices.dices)
	{
		if (dice == 6)
		{
			score += dice;
		}
	}
	return score;
}

int Judger::Chance(const Dices& dices)
{
	int score = 0;
	for (int dice : dices.dices)
	{
		score += dice;
	}
	return score;
}

int Judger::ThreeOfAKind(const Dices& dices)
{
	std::map<int, int> m;
	for (int dice : dices.dices)
	{
		m[dice]++;
	}

	for (auto count : m)
	{
		if (count.second >= 3)
		{
			int sum = 0;
			for (int dice : dices.dices)
			{
				sum += dice;
			}
		}
	}
	return 0;
}

int Judger::FourOfAKind(const Dices& dices)
{
	std::map<int, int> m;
	for (int dice : dices.dices)
	{
		m[dice]++;
	}

	for (auto count : m)
	{
		if (count.second >= 4)
		{
			int sum = 0;
			for (int dice : dices.dices)
			{
				sum += dice;
			}
		}
	}
	return 0;
}

int Judger::FullHouse(const Dices& dices)
{
	std::map<int, int> m;
	for (int dice : dices.dices)
	{
		m[dice]++;
	}

	if (m.size() == 2 && m.begin()->second == 2 || m.begin()->second == 3)
	{
		return 25;
	}
	return 0;
}

int Judger::SmallStraight(const Dices& dices)
{
	std::set<int> s;
	for (int dice : dices.dices)
	{
		s.insert(dice);
	}
	std::vector<std::set<int>> prefabs =
	{
		{1, 2, 3, 4},
		{2, 3, 4, 5},
		{3, 4, 5, 6}
	};
	for (std::set<int>& prefab : prefabs)
	{
		bool matched = true;
		for (int dice : prefab)
		{
			if (s.find(dice) == s.end())
			{
				matched = false;
				break;
			}
		}

		if (matched)
		{
			return 30;
		}
	}
	return 0;
}

int Judger::LargeStraight(const Dices& dices)
{
	std::set<int> s;
	for (int dice : dices.dices)
	{
		s.insert(dice);
	}
	std::vector<std::set<int>> prefabs =
	{
		{1, 2, 3, 4, 5},
		{2, 3, 4, 5, 6},
	};
	for (std::set<int>& prefab : prefabs)
	{
		bool matched = true;
		for (int dice : prefab)
		{
			if (s.find(dice) == s.end())
			{
				matched = false;
				break;
			}
		}

		if (matched)
		{
			return 40;
		}
	}
	return 0;
}

int Judger::Yahtzee(const Dices& dices)
{
	std::map<int, int> m;
	for (int dice : dices.dices)
	{
		m[dice]++;
	}

	if (m.begin()->second == 5)
	{
		return 50;
	}

	return 0;
}
